patches/net/minecraft/network/codec/ByteBufCodecs.java.patch

```diff
@@ -13,3 +13,16 @@
              }
  
              public R decode(RegistryFriendlyByteBuf p_331253_) {
+@@ -493,7 +_,11 @@
+             private static final int DIRECT_HOLDER_ID = 0;
+ 
+             private IdMap<Holder<T>> getRegistryOrThrow(RegistryFriendlyByteBuf p_341377_) {
+-                return p_341377_.registryAccess().registryOrThrow(p_320391_).asHolderIdMap();
++                var registry = p_341377_.registryAccess().registryOrThrow(p_320391_);
++                if (net.neoforged.neoforge.registries.RegistryManager.isNonSyncedBuiltInRegistry(registry)) {
++                    throw new IllegalStateException("Cannot use ID syncing for non-synced built-in registry: " + registry.key());
++                }
++                return registry.asHolderIdMap();
```

Comment:
Instead of duplicating this code between here and the `getRegistryOrThrow` method in `registry(ResourceKey, Function)`,  I think it would make more sense to add a hook method that basically is:

```java
public static <T> Registry<T> getSyncableRegistryOrThrow(RegistryFriendlyByteBuf buffer, ResourceKey<? extends Registry<T>> registryKey) {
    Registry<T> registry = buffer.registryAccess().registryOrThrow(registryKey);
    if (RegistryManager.isNonSyncedBuiltInRegistry(registry)) {
        throw new CodecException("Cannot use ID syncing for non-synced built-in registry: " + registry.key());
    }
    return registry;
}
```
(CodecException being from io.netty.handler.codec)

And then this patch instead becomes:
```diff
- return p_341377_.registryAccess().registryOrThrow(p_320391_).asHolderIdMap();
+ return  hooks.getSyncableRegistryOrThrow(p_341377_,  p_320391_).asHolderIdMap();
```
And the diff in the existing spot would become:
```diff
- return p_320353_.apply(p_330361_.registryAccess().registryOrThrow(p_319942_));
+ return p_320353_.apply(hooks.getSyncableRegistryOrThrow(p_330361_,  p_319942_));
```